lib/vquic/quiche.c

```diff
@@ -831,14 +877,27 @@ CURLcode Curl_quic_done_sending(struct connectdata *conn)
   if(conn->handler == &Curl_handler_http3) {
     /* only for HTTP/3 transfers */
     ssize_t sent;
+    int i;
+    curl_socket_t sockfd;
     struct HTTP *stream = conn->data->req.p.http;
     struct quicsocket *qs = conn->quic;
-    fprintf(stderr, "!!! Curl_quic_done_sending\n");
     stream->upload_done = TRUE;
     sent = quiche_h3_send_body(qs->h3c, qs->conn, stream->stream3_id,
                                NULL, 0, TRUE);
-    if(sent < 0)
-      return CURLE_SEND_ERROR;
+    if(sent < 0) {
+      if(sent == QUICHE_H3_ERR_DONE) {
+        return CURLE_AGAIN;
```

Comment:
This quiche error is "documented" (in the header) as `There is no more work to do.`. Why do we need to EAGAIN it?

Reply:
@bagder I do this because I read the source code of quiche:
```rust
        if stream_cap <= overhead {
            return Err(Error::Done);
        }
```
url: https://github.com/cloudflare/quiche/blob/master/src/h3/mod.rs#L1026 

And the document of this method mentioned:

```
    /// When a partial write happens (including when [`Done`] is returned) the
    /// application should retry the operation once the stream is reported as
    /// writable again.
```

When you upload a big file, the quiche send buffer could be exhausted, and the `stream_capacity`  here could be zero, but the `overhead` will never be zero, so this method `send_body()` would returns `Done`. In this situation, curl should retry later, rather than return `CURLE_SEND_ERROR`.

Reply:
I think that fairly unusual mandate could warrant a comment next to the code I asked about so that the next fellow developer reading that part of the code can get an explanation for it as well!